[OH3] Migrate to either OH3 or OH2 incl. existing files on disk

mstormi opened the pull request:
Installs new, removes old packages.
Changes filenames and config on disk to work with either openHAB3 or master/stable branches.
Must be part of master because on migrating (back) *to* either master or stable, all openHABian code is re-downloaded from that target branch *and script will restart* with anything overwritten we had put into the openHAB3 branch.

Signed-off-by: Markus Storm <markus.storm@gmx.net>

ecdye reviewed (commented):
I need to look into this more to make sure that this is everything but LGTM.

mstormi commented on functions/openhabian.bash:
```diff
@@ -134,10 +166,10 @@ openhabian_update() {
   if [[ -n $INTERACTIVE ]]; then
-    if [[ $current == "stable" || $current == "master" ]]; then
-      if ! selection="$(whiptail --title "openHABian version" --radiolist "$introText" 14 80 2 stable "recommended standard version of openHABian" ON master "very latest version of openHABian" OFF 3>&1 1>&2 2>&3)"; then return 0; fi
+    if [[ $current == "stable" || $current == "master" || $current == "openHAB3" ]]; then
```
See my introductory comment. Did that in my earlier version but trouble is on downgrade the script downloads master or stable and the migration code is missing there.

mstormi commented:
> I need to look into this more to make sure that this is everything but LGTM.

I had forgotten to call create_user if user or group doesn't exist. Now check and eventually add in openhabian-config on every start. Is there a better location to check you can think of ?

ecdye commented:
@mstormi Does this migrate configs too? Or do we expect the user to do that?

mstormi commented:
> Does this migrate configs too?

No. I believe if you remove the openhab2 package it leaves the config behind (on purge, too?) and the install scripts from the OH3 package will offer to keep or overwrite every file that changes. (@BClark09 correct ?).
But there's no downgrade script to do the inverse. So all we can do is ex+reimport the config.

User needs to do that currently. I'll expand restore_openhab_config() to work with files in both backupdirs, the OH2 and OH3 ones. I was thinking to also automate that step but I'd be happier to have a _working_ openHABian codepiece first.

mstormi commented:
I'm really having problems with integration testing this because on OH2<->OH3 branch changes, new to-be-tested code is replaced from the "other" old repo.
Best I can think of ATM is to merge #1206 first and then this, then do a full test and identify what's left to work on.
@ecdye Wdyt?

ecdye commented:
Should be fine, the openHAB3 branch is not being used in any production environments and it is to be expected if things break over there.

mstormi commented:
ok merged to openHAB3.
OH3 needs Java 11 and complains if that's not installed. We should do a forced upgrade, what would be the proper code snippet for that?

ecdye commented:
> ok merged to openHAB3.
> OH3 needs Java 11 and complains if that's not installed. We should do a forced upgrade, what would be the proper code snippet for that?

Probably the best way to do that would be to display a notice on startup and ask if the user approves, then if so we just call `java_install_or_update "Zulu11-32"` and update `/etc/openhabian.conf`

mstormi commented:
This is d*mned complex to work with,always have to patch both branches, prepMigration and openHAB3 but cannot use the same codebase.

mstormi commented:
> Probably the best way to do that would be to display a notice on startup and ask if the user approves, then if so we just call `java_install_or_update "Zulu11-32"` and update `/etc/openhabian.conf`

This is also on interactive install and a user might have Java 11 running already (Zulu or even Adopt).
What's the most easy check for a compatible (11) Java ? Will java_install_or_update "Zulu11-32" overwrite that ?

mstormi commented:
> java_install_or_update "Zulu11-32"

I inserted that in my latest code. Now it *downloads* that. Shouldn't it be in the cache ? Or is that just Java 8 ?

mstormi commented:
I believe it's working now but I still cannot do an integration test of a downgrade, would need to merge to master first.
So at this point I think we need to merge and see what's left to not work.

Could you please review now ?

mstormi commented:
Hmm, a fresh install from image downloads a wrong branch because of
.git/config:    url = https://github.com/mstormi/openhabian
... and I cannot find the reason for that file to exist.

mstormi commented:
any feedback, @ecdye @holgerfriedrich @BClark09 anyone ?

ecdye commented:
> > java_install_or_update "Zulu11-32"
> 
> I inserted that in my latest code. Now it _downloads_ that. Shouldn't it be in the cache ? Or is that just Java 8 ?

Just Java 8 unless it was on 64 bit. We had kept it until OH3 in case of compatibility issues as it was stable for certain on Java 8.

mstormi commented:
prefetch Zulu11-32 please. In case of incompatibilities a 2.5 user can go back and install Zulu8. An OH3 user cannot anyway.

ecdye commented:
Seems to be working, needs some polish for cross branch changes but overall seems to be working!

mstormi commented:
> I had to make a few changes to the migration code

which changes, [these ?](https://github.com/openhab/openhabian/commit/291c87d89213c2eedd8eb2c9986f404a84aaf73e#diff-2b13d353215584eb937ab19a145d01b10448f3abe4119e520f2f23be8aae5bb7) 

Updated my PR, please cross-check if that's fine with you now.

AFAIK there's a means for a reviewer in Github to suggest changes to PR that the author can simply accept. A way safer method than a message like this one. Must admit though I have not figured out myself how that works but maybe you know or find out.

ecdye commented on docs/openhabian.md:
```diff
@@ -290,6 +224,97 @@ This is not done by default because it requires a number of user inputs, but you
+
+Mind the comments of a configuration parameter. Browse the next documentation section for further explanations.
+
+#### administration user
```
Capitalize administration

ecdye commented on docs/openhabian.md:
```diff
@@ -290,6 +224,97 @@ This is not done by default because it requires a number of user inputs, but you
 
+## Setup notes
+
+### on openHAB3
```
Capitalize on

ecdye commented on functions/backup.bash:
```diff
@@ -44,15 +42,15 @@ restore_openhab_config() {
 
   local backupList
-  local backupPath="/var/lib/openhab2/backups"
+  local backupPath="${OPENHAB_BACKUPS:-/var/lib/openhab2/backups}"
```
Where does this come from, meaning `$OPENHAB_BACKUPS`

mstormi replied:
/etc/default/openhab[2]
but only if manually specified there which it isn't by default.

ecdye commented on functions/menu.bash:
```diff
@@ -25,12 +25,14 @@ Note backups are NOT active per default so remember to set them up right at the
   "01 | Select Branch"           "Select the openHABian config tool version (\"branch\") to run" \
   "02 | Upgrade System"          "Upgrade all installed software packages (incl. openHAB) to their latest version" \
   "03 | openHAB Stable"          "Install or upgrade to the latest stable release of openHAB 2" \
+  "04 | Upgrade to openHAB 3"    "Upgrade to openHAB 3 testing milestone (BETA !)" \
```
Should we force switch to OH3 Branch when this option is selected?

mstormi replied:
I was wondering, too.
I migrate when the user selects to switch to openHAB3 branch but don't when he switches from there elsewhere (back down), that's how the upgrade was meant to be triggered. But we had no downgrade path (as we cannot be sure if a specific branch is OH 2 or 3) so I introduced options 04 and 05
On 2nd thought yes we must switch branch or else future actions triggered from the tool will fail.

ecdye commented on functions/openhab.bash:
```diff
@@ -182,13 +189,29 @@ multitail_openhab_scheme() {
+##    openhab2_is_installed()
+##
+openhab2_is_installed() {
+  if [[ $(dpkg -s 'openhab2' | grep Status | cut -d' ' -f2) == "install" ]]; then return 0; else return 1; fi
```
use the `dpkg -s 'openhab2' &> /dev/null` style to avoid unwanted output on the user side.

mstormi replied:
ok

ecdye commented on includes/adminkey.pub:
```diff
@@ -0,0 +1 @@
+ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDeEZqjLjbZmqFrhYL1inJM/U92m1/rZ+Lq/ZJvsEITk0I/7GBJqZ5Ze4Oyh/XKtbCC3l9QTwfhOyd/jlcjIaKaRPznY0Prml11lMY6OJHLPaBvKISaRoZb/J9xk0fwRAuXoVQNKMCE5DXCIBYNWEILex6Z2MzESRwbsEciDxdIuBSsJA9IPfCrFZhrgm/oBCtvEKflQCwxMcBI3kdfPUInFpqpafvvNJl0uIexsRMrKEyUMJ7+WDVXc/UQm4P/dodiLPXZkppwrPg5OzqIZWW5TZglCtlo5hMOMD4fTyKplTwWXliKtrGrb54lELCHJpSGDhx/y9aeX+mLXszXeC/QFWPg3DWaFpbaix6NGTR ...
```
This should not be a change included in the master branch, right? Unless you want your key publicly available.

mstormi replied:
right, this was for testing only, will remove

ecdye reviewed (changes requested):
Looks pretty good, just a few changes needed. When you do merge please be sure that the code is consistent throughout all branches.

ecdye commented:
> AFAIK there's a means for a reviewer in Github to suggest changes to a PR that the author can simply accept. A way safer method than a message like this one. Must admit though I have not figured out myself how that works but maybe you know or find out.

Hmm, I actually did not know that. I know that a reviewer can just make changes if the PR author selects the box that allows it but not a nice way like that. I'll look into it because that would be very convenient.

mstormi commented:
Synced changes with openHAB3 again, included your changes and added switching branches on both up AND downgrades as the tool needs to work on the changed branch (so it needs to be replaced/restarted via branch switch).
Untested though. Merge like this ?

ecdye commented:
Is there a way to hide the menu options a little bit more so that we don't advertise potentially broken functionality?

mstormi commented:
> Is there a way to hide the menu options a little bit more so that we don't blatantly advertise potentially broken functionality?

I moved them to the 40 menu
Also rebased openHAB3 on master

Please have a final look before I merge

mstormi commented:
Noticed that Cancel on install and the check for an already installed OH didn't work so reordered some.
Seems to work now, hope I didn't overlook and break any of the other use cases.
Updated NEWS, too.
Will merge now and put up a small announcement.
